### PR TITLE
Don't deadlock with SyncStream

### PR DESCRIPTION
**Pull request by James-Ko**
The following scenario will cause a deadlock in the old API because `Stream.Synchronized()` locks on the underlying Stream:

    Stream stream = MakeAStream(); // i.e. Stream.Null
    Stream synchronizedStream = Stream.Synchronized(stream);
    lock (stream)
    {
        var thread = new Thread(() =>
        {
            long length = synchronizedStream.Length; // deadlock here because synchronizedStream internally locks on stream
        });
    
        thread.Start();
        thread.Join();
    }

To fix this, the `SyncStream` class that `Stream.Synchronized()` returns  has been made to lock on a NonSerialized object instead of the Stream itself.

### CODE REVIEW

**Comment by hadibrais** [comment]
@James-Ko  But why would you lock the stream object if thread is going to use the synchronized stream object? You should just use the synchronized stream. An additional lock is only required when you want to synch access to the stream across appdomains or processes. Even in this case, locking the steam object won't do the job. Introducing another field is not necessary.

**Comment by James-Ko** [comment]
@hadibrais True, but that was only an example to show the shortest repro case; there may be other, more obscure cases (i.e. the first lock is acquired through another library). Also, **how can the user of `Stream.Synchronized()` be expected to know he can't lock on the base Stream,** especially when [the documentation says nothing about it](https://msdn.microsoft.com/en-us/library/system.io.stream.synchronized(v=vs.110).aspx)? While there may not be many cases for keeping the Stream around after synchronization, it is not illegal to.

File: src/mscorlib/src/System/IO/Stream.cs [hunk 1]
```diff
@@ -1059,6 +1059,8 @@ internal sealed class SyncStream : Stream, IDisposable
             private bool? _overridesBeginRead;
             [NonSerialized]
             private bool? _overridesBeginWrite;
+            [NonSerialized]
+            private readonly object _syncRoot = new object();
```
Comment (stephentoub) [inline src/mscorlib/src/System/IO/Stream.cs]: With this marked [NonSerialized], won't _syncRoot be left uninitialized and null when such a SyncStream is deserialized, causing NullReferenceExceptions on subsequent use?

**Comment by stephentoub** [comment]
While I definitely agree with you that it'd have been better if from the start Stream.Synchronized didn't lock on the underlying stream instance, I do worry about compat issues a change like this could cause.  Someone today could be locking on their stream instance and expecting such operations to be synchronized with operations on the wrapping SyncStream, or they could be handing out multiple instances of Stream.Synchronized for the same underlying stream and expecting all wrapping streams to be synchronized with each other, and this change will cause such assumptions to fail, resulting in the potential for silent race conditions to be introduced into existing programs.

To my knowledge, this behavior has been this way since v1, and I've not heard of this issue being a real problem (I agree it could be in theory).  Unless proven otherwise, I'm inclined to think the best course of action is to leave this as-is.

**Pull request closed without merging**